### ISSUE TITLE
`LearningRateMonitor` doesn't reflect `LearningRateFinder` changes

### ISSUE THREAD

**Issue by davidgilbertson** [Lightning-AI/lightning#15856]
### Bug description

It seems that when `LearningRateFinder` changes the LR of my model's optimizer, this isn't reflected in the values logged by `LearningRateMonitor`.

I think the offending line of code is here: https://github.com/Lightning-AI/lightning/blob/master/src/pytorch_lightning/callbacks/lr_monitor.py#L187-L191 as at version 1.8.3.post1

Where `trainer.optimizers` returns the optimizer with the original LR value, but `trainer.model.optimizers()` has the updated value.

### Error messages and logs

```

# Error messages and logs here please

```

### Environment

```

#- Lightning Component (e.g. Trainer, LightningModule, LightningApp, LightningWork, LightningFlow):
#- PyTorch Lightning Version (e.g., 1.5.0):
#- Lightning App Version (e.g., 0.5.2):
#- PyTorch Version (e.g., 1.10):
#- Python version (e.g., 3.9):
#- OS (e.g., Linux):
#- CUDA/cuDNN version:
#- GPU models and configuration:
#- How you installed Lightning(`conda`, `pip`, source):
#- Running environment of LightningApp (e.g. local, cloud):

```

**Comment by SkafteNicki** [comment]
Not sure when this issue was fixed but currently on master I am unable to reproduce the issue. Below is a small reproducible script that showcases that the learning rate is correctly updated after running learning rate finder and that value is the logged to the learning rate monitor.

```python
import torch
import torch.nn as nn
import torch.nn.functional as F
from torch.utils.data import DataLoader, Dataset

import lightning.pytorch as pl
from lightning.pytorch.callbacks import LearningRateMonitor, LearningRateFinder
from lightning.pytorch.loggers import CSVLogger


class SimpleDataset(Dataset):
    """Simple dataset for testing purposes."""
    
    def __init__(self, size=100, input_dim=10):
        self.data = torch.randn(size, input_dim)
        self.targets = torch.randn(size, 1)
    
    def __len__(self):
        return len(self.data)
    
    def __getitem__(self, idx):
        return self.data[idx], self.targets[idx]


class SimpleModel(pl.LightningModule):
    """Simple model for testing LR monitoring and finding."""
    
    def __init__(self, input_dim=10, lr=1e-3):
        super().__init__()
        self.save_hyperparameters()
        
        self.layer1 = nn.Linear(input_dim, 64)
        self.layer2 = nn.Linear(64, 32)
        self.layer3 = nn.Linear(32, 1)
        self.dropout = nn.Dropout(0.1)
        
        # Store initial learning rate
        self.lr = lr
    
    def forward(self, x):
        x = F.relu(self.layer1(x))
        x = self.dropout(x)
        x = F.relu(self.layer2(x))
        x = self.dropout(x)
        return self.layer3(x)
    
    def training_step(self, batch, batch_idx):
        x, y = batch
        y_hat = self(x)
        loss = F.mse_loss(y_hat, y)
        
        # Log the current learning rate from optimizer
        current_lr = self.optimizers().param_groups[0]['lr']
        self.log('train_loss', loss, prog_bar=True)
        self.log('current_lr_from_optimizer', current_lr, prog_bar=True)
        
        return loss
    
    def validation_step(self, batch, batch_idx):
        x, y = batch
        y_hat = self(x)
        loss = F.mse_loss(y_hat, y)
        self.log('val_loss', loss, prog_bar=True)
        return loss
    
    def configure_optimizers(self):
        optimizer = torch.optim.Adam(self.parameters(), lr=self.lr)
        return optimizer
    
    def train_dataloader(self):
        dataset = SimpleDataset(size=200, input_dim=10)
        return DataLoader(dataset, batch_size=16, shuffle=True)
    
    def val_dataloader(self):
        dataset = SimpleDataset(size=50, input_dim=10)
        return DataLoader(dataset, batch_size=16, shuffle=False)


class CustomLearningRateFinder(LearningRateFinder):
    """Extended LearningRateFinder that provides more detailed logging."""
    
    def on_fit_start(self, trainer, pl_module):
        print(f"  LR Finder starting...")
        print(f"    Initial LR from model.lr: {pl_module.lr}")
        print(f"    Initial LR from optimizer: {trainer.optimizers[0].param_groups[0]['lr']}")
        
        super().on_fit_start(trainer, pl_module)
        
        if self.optimal_lr:
            print(f"    Suggested optimal LR: {self.optimal_lr.suggestion()}")
            print(f"    Final LR in optimizer: {trainer.optimizers[0].param_groups[0]['lr']}")


def main():
    model = SimpleModel(lr=1e-4)
    print(f"Initial model LR: {model.lr}")
    
    # Create LearningRateFinder and LearningRateMonitor
    lr_finder = CustomLearningRateFinder(
        min_lr=1e-6,
        max_lr=1e-1,
        num_training_steps=50,
        early_stop_threshold=4.0,
        update_attr=True
    )
    lr_monitor = LearningRateMonitor(logging_interval='step')
    
    # Create trainer with both callbacks
    trainer = pl.Trainer(
        max_epochs=2,
        limit_train_batches=10,
        limit_val_batches=3,
        callbacks=[lr_finder, lr_monitor],
        logger=CSVLogger("./logs", name="issue_15856_test"),
        enable_checkpointing=False,
        enable_progress_bar=False,  # Keep output clean
        log_every_n_steps=1
    )
    
    print("\n" + "="*50)
    print("BEFORE TRAINING:")
    print(f"  Model lr attribute: {model.lr}")
    print("="*50)
    
    # Train the model
    trainer.fit(model)
    
    print("\n" + "="*50)
    print("AFTER LR FINDER & TRAINING:")
    print(f"  Model lr attribute: {model.lr}")
    print(f"  Optimizer lr: {trainer.optimizers[0].param_groups[0]['lr']}")
    
    # The key test: Check if LearningRateMonitor logged the updated LRs
    print(f"\n  LearningRateMonitor results:")
    for key, values in lr_monitor.lrs.items():
        if values:
            print(f"    {key}: Logged {len(values)} values")
            print(f"      First LR: {values[0]}")
            print(f"      Last LR: {values[-1]}")

if __name__ == "__main__":
    main()
```
Closing issue, but feel free to ping and reopen if necessary.
